### ISSUE TITLE
Bug: `useOptimistic` reducer is called twice (not in strict-mode)

### ISSUE THREAD

**Issue by arklanq** [facebook/react#27702]
#### Some context first:

I was reading about the new ["useOptimistic" hook in the official documentation](https://react.dev/reference/react/useOptimistic) and came across the provided example. I have forked this example on CodeSandbox and played a little with some console.log statements in order to fully understand the feature. While I was debugging I found something that truly looks like a bug to me.

#### The problem
The `useOptimistic` reducer is run twice:
- 1st time when intentionally invoked by the programmer (✅);
- 2d time when the original state changes (❓).

This behavior leads to an undesirable state of the component. In the provided example:  `optimisticMessages` array hold for a moment as many as 3 items, which 2 of them are duplicates of optimistically updated item. 

I am not sure if this invalid state lasts only for a few milliseconds or is it a separate concurrent render with different state. If the latter, this should be explained in the docs.

#### React version:

18.3.0-canary-6db7f4209-20231021

## Steps To Reproduce

1. See the docs, get familiar with an example:
https://react.dev/reference/react/useOptimistic

2. Go to my fork and check out an extra debugging statements I have implemented. I have also disabled strict mode in order to get clear logs without duplicates.
https://codesandbox.io/p/sandbox/react-canary-useoptimistic-jmfhhg

3. Observe the console logs.

![image](https://github.com/facebook/react/assets/23725071/2d0684de-fcf7-442c-8e2f-54293f941b17)

Link to code example:

https://codesandbox.io/p/sandbox/react-canary-useoptimistic-jmfhhg

## The current behavior

`useOptimistic` hook runs twice, resulting in inconsistent state.

## The expected behavior

- `useOptimistic` reducer should run only once;
- ... or it should be better explained in the docs, why the reducer is run each time the original state changes and how to deal with it.

**Comment by manivsh** [comment]
Also why isn't the `useOptimistic()` tracking the state of the transaction? and just take the change of 1st parameter as `done`? 

Something like this (👇) is out of scope? 

```typescript 
[messages, status, doTransaction] = useOptimistic(propMessages, f1, f2);

//messages is the state based on which the render shoud depend.
//status could be the status of the Promise

//f1 is a pure function for calculating temporary state 
//f2 is the actual function that mutate data, and return a Promise<message>.

```

**Comment by nctay** [comment]
In case anyone wondering, this is still an issue in current 19rc. 
Passing reducer to useOptimistic hook causes it triggers twice. 
Do not pass reducer, update state with function that returns from hook:

```
 const [optimisticState, setOptimisticState] = useOptimistic(initialData);
 
 const updater = (newData) => {
    setOptimisticState(newData) 
 }
```

**Comment by eps1lon** [comment]
Duplicate of https://github.com/facebook/react/issues/28574
